### ISSUE TITLE
ClassLoader constraint issue when using NativeWebSocketConfiguration with WEB-INF/lib/jetty-http.jar present

### ISSUE THREAD

**Issue by joakime** [eclipse/jetty.project#1229]
As reported in jetty-dev by Michele Rossi
https://dev.eclipse.org/mhonarc/lists/jetty-dev/msg02862.html

> I am getting the following exception when trying to start up the WebSockets transport using CometD 3.1.0 and Jetty 9.3.15:
> 
> 
> 14:39:16.244|WARN |o.e.j.d.DeploymentManager||Unable to reach node goal: started
> java.lang.LinkageError: loader constraint violation: when resolving method "org.eclipse.jetty.websocket.server.NativeWebSocketConfiguration.addMapping(Lorg/eclipse/jetty/http/pathmap/PathSpec;Lorg/eclipse/jetty/websocket/servlet/WebSocketCreator;)V" the class loader (instance of org/eclipse/jetty/webapp/WebAppClassLoader) of the current class, org/cometd/websocket/server/JettyWebSocketTransport, and the class loader (instance of org/eclipse/jetty/start/Classpath$Loader) for the method's defining class, org/eclipse/jetty/websocket/server/NativeWebSocketConfiguration, have different Class objects for the type org/eclipse/jetty/http/pathmap/PathSpec used in the signature
>         at org.cometd.websocket.server.JettyWebSocketTransport.init(JettyWebSocketTransport.java:75)
>         at org.cometd.server.BayeuxServerImpl.initializeServerTransports(BayeuxServerImpl.java:255)
>         at org.cometd.server.BayeuxServerImpl.doStart(BayeuxServerImpl.java:135)
>         at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
>         at org.cometd.server.CometDServlet.init(CometDServlet.java:64)

Need to introduce testcase where the WEB-INF/lib/jetty-http.jar is present in a WebAppContext based deployment (WAR file).

**Comment by joakime** [comment]
Junit testcase for this does not trigger the classloader constraint.

Testing with a manually deployed WAR file next.

**Comment by joakime** [comment]
Was able to reproduce with Jetty 9.3.15 and the attached WAR (compressed with gzip to get around github file upload restrictions).

[ws-test.war.gz](https://github.com/eclipse/jetty.project/files/696500/ws-test.war.gz)

```
2017-01-10 11:19:46.541:WARN:oejd.DeploymentManager:main: Unable to reach node goal: started
java.lang.LinkageError: loader constraint violation: when resolving method
 "org.eclipse.jetty.websocket.server.NativeWebSocketConfiguration.addMapping(Lorg/eclipse/jetty/http/pathmap/PathSpec;Lorg/eclipse/jetty/websocket/servlet/WebSocketCreator;)V"
 the class loader (instance of org/eclipse/jetty/webapp/WebAppClassLoader) of the current class, org/eclipse/jetty/websocket/server/InfoContextAttributeListener,
 and the class loader (instance of org/eclipse/jetty/start/Classpath$Loader)
 for the method's defining class, org/eclipse/jetty/websocket/server/NativeWebSocketConfiguration, have different Class objects for the type org/eclipse/jetty/http/pathmap/PathSpec used in the signature
	at org.eclipse.jetty.websocket.server.InfoContextAttributeListener.contextInitialized(InfoContextAttributeListener.java:36)
	at org.eclipse.jetty.server.handler.ContextHandler.callContextInitialized(ContextHandler.java:843)
	at org.eclipse.jetty.servlet.ServletContextHandler.callContextInitialized(ServletContextHandler.java:533)
	at org.eclipse.jetty.server.handler.ContextHandler.startContext(ContextHandler.java:816)
	at org.eclipse.jetty.servlet.ServletContextHandler.startContext(ServletContextHandler.java:345)
	at org.eclipse.jetty.webapp.WebAppContext.startWebapp(WebAppContext.java:1404)
	at org.eclipse.jetty.webapp.WebAppContext.startContext(WebAppContext.java:1366)
	at org.eclipse.jetty.server.handler.ContextHandler.doStart(ContextHandler.java:778)
	at org.eclipse.jetty.servlet.ServletContextHandler.doStart(ServletContextHandler.java:262)
	at org.eclipse.jetty.webapp.WebAppContext.doStart(WebAppContext.java:520)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
	at org.eclipse.jetty.deploy.bindings.StandardStarter.processBinding(StandardStarter.java:41)
	at org.eclipse.jetty.deploy.AppLifeCycle.runBindings(AppLifeCycle.java:188)
	at org.eclipse.jetty.deploy.DeploymentManager.requestAppGoal(DeploymentManager.java:499)
	at org.eclipse.jetty.deploy.DeploymentManager.addApp(DeploymentManager.java:147)
	at org.eclipse.jetty.deploy.providers.ScanningAppProvider.fileAdded(ScanningAppProvider.java:180)
	at org.eclipse.jetty.deploy.providers.WebAppProvider.fileAdded(WebAppProvider.java:452)
	at org.eclipse.jetty.deploy.providers.ScanningAppProvider$1.fileAdded(ScanningAppProvider.java:64)```

**Comment by joakime** [comment]
I can get past the classloader constraint with a system/server classloader tweak.

``` xml
<?xml version="1.0"?>
<!DOCTYPE Configure PUBLIC "-//Jetty//Configure//EN" "http://www.eclipse.org/jetty/configure_9_3.dtd">

<Configure class="org.eclipse.jetty.webapp.WebAppContext">
  <Set name="contextPath">/ws-test</Set>
  <Set name="war"><Property name="jetty.webapps"/>/ws-test.war</Set>
  <Call name="prependSystemClass">
    <Arg>org.eclipse.jetty.http.pathmap.</Arg>
  </Call>
  <Call name="prependServerClass">
    <Arg>-org.eclipse.jetty.http.pathmap.</Arg>
  </Call>
</Configure>
```

@sbordet and @gregw do we want to set the above 2 entries in the default WebAppContext/WebAppClassloader configuration?

But I hit another bug in the 9.3.15 release ...

```
2017-01-10 11:37:20.428:WARN:oejw.WebAppContext:main: Failed startup of context o.e.j.w.WebAppContext@4883b407{/ws-test,file:///private/var/folders/w5/mmnzpk0n369dntp4nszlc8h40000gn/T/jetty-0.0.0.0-8080-ws-test.war-_ws-test-any-5892397045606348966.dir/webapp/,UNAVAILABLE}{/ws-test.war}
javax.servlet.ServletException: org.eclipse.jetty.websocket.server.WebSocketUpgradeFilter is defined twice for the same context attribute key 'org.eclipse.jetty.websocket.server.WebSocketUpgradeFilter'.  Make sure you have different init-param 'contextAttributeKey' values set
	at org.eclipse.jetty.websocket.server.WebSocketUpgradeFilter.setToAttribute(WebSocketUpgradeFilter.java:380)
	at org.eclipse.jetty.websocket.server.WebSocketUpgradeFilter.init(WebSocketUpgradeFilter.java:354)
	at org.eclipse.jetty.servlet.FilterHolder.initialize(FilterHolder.java:139)
	at org.eclipse.jetty.servlet.ServletHandler.initialize(ServletHandler.java:873)
	at org.eclipse.jetty.servlet.ServletContextHandler.startContext(ServletContextHandler.java:349)
	at org.eclipse.jetty.webapp.WebAppContext.startWebapp(WebAppContext.java:1404)
	at org.eclipse.jetty.webapp.WebAppContext.startContext(WebAppContext.java:1366)
```

Which was fixed in commit 0fc8972, which will require another release of Jetty 9.3.x

**Comment by sbordet** [comment]
Requires to be backported to 9.2.x as well.

**Comment by sbordet** [comment]
Talking with @gregw the solution of punching another hole in the server classes is not good.

**Comment by gregw** [comment]
while is it acceptable to punch a hole in the classloader for something like MySpecialPurposeFilter,  it is not OK to do so for utility classes like PathMap.   If we did, then any app that uses the jetty client would be constrained to only run on a version of the server with a compatible PathMap instance, which could constrain how we develop that class over time.

What I'm not understanding is why the `InfoContextAttributeListener` is seeing the webapps version of `PathSpec` in the first place?   `InfoContextAttributeListener` should be on the container classpath as a system class that cannot be overriden, thus it should be loaded by the jetty container classloader and should see the containers `PathSpec` class instance.

**Comment by joakime** [comment]
The `InfoContextAttributeListener` is a `ServletContextListener` that configures the WebSocket endpoints for that specific webapp.

Code for InfoContextAttributeListener can be found at 
https://github.com/eclipse/jetty.project/blob/jetty-9.3.x/jetty-websocket/websocket-server/src/test/java/org/eclipse/jetty/websocket/server/InfoContextAttributeListener.java

This concept, having a ServletContextListener configure the WebSocket endpoints, is rather common, found in Spring and Cometd (for example).

Perhaps the API for NativeWebSocketConfiguration could be string based, but uses the internal jetty-http PathSpec classes without exposing them on the API.  But that will cause a ripple effect down to the backward compatibility with the WebSocketUpgradeFilter too, making the new version not API compatible with older WebSocketUpgradeFilter usages.

One note (but unrelated to this specific issue): The WebSocket implementation on the JSR side is tied to the Server based jetty-client classes (JSR makes Client the base class for the Server, meaning that Client is mandatory for a Server), having WebApp specific jetty-client libs is going to be a problem with JSR356. (probably true for WebSocket native client now too, as we've integrated jetty-client with WebSocketClient for the proxy / auth support)

**Comment by gregw** [comment]
Oh I see, the InfoContextAttributeListener is just a test class. So it can't be loaded from the container classpath!

Well that is indeed the issue with native APIs - they just can't be called easily from code that is loaded from the web application .

So I think the `NativeWebSocketConfiguration` API does have to be based on Strings (or at least on jetty utility classes).   I hear you about t he ripple effect, but it may be that  you need to have two versions of it, one for embedded usage and one for webapp usage.

I like that  you are getting the instance via an fake attribute, so that is good isolation, but it needs to avoid classes in the API as well.

**Comment by joakime** [comment]
Here's the latest webapp to test with.
[ws-test.war.gz](https://github.com/eclipse/jetty.project/files/709443/ws-test.war.gz)

The `issue-1229` branch is being used to complete this issue, however it has revealed an interesting regression (filed as #1264) in `ServletContainerInitializer` behavior in Jetty 9.3.x

**Comment by sbordet** [comment]
Confirmed that CometD works fine with the changes in branch `issue-1229`.

**Comment by gregw** [comment]
@joakime  should we merge this branch so we can include it in testing while we get ready for a 9.4.1?.  Don't want last minute merges

**Comment by sbordet** [comment]
@gregw I reviewed with @joakime and will merge tomorrow my morning. Needs to be merged from 9.2.x onwards.

**Comment by sbordet** [comment]
Branch `issue-1229` has been merged into branch `jetty-9.2.x` and merged all the way forward.

I had to remove the JDK8-isms for 9.2.x, cleaned up a bit the implementation by removing code duplication, guarded against NPEs in a case, but from initial testing looks the new code fixes the issue.

**Comment by joakime** [comment]
@sbordet has merged, this is closed
